onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp

```diff
@@ -234,60 +234,66 @@ DML_OP_EXTERN_QUERY_FUNCTION(MaxPool);
-const static SupportedTensorDataTypes supportedTypeListScatterGatherND[1] = { SupportedTensorDataTypes::Scalars8to32 };
-const static SupportedTensorDataTypes supportedTypeListSlice10[2] = { SupportedTensorDataTypes::NumericDefault, SupportedTensorDataTypes::Int32 | SupportedTensorDataTypes::Int64 };
-const static SupportedTensorDataTypes supportedTypeListQuantizeLinear[2] = { SupportedTensorDataTypes::Float32 | SupportedTensorDataTypes::Int32, SupportedTensorDataTypes::UInt8 | SupportedTensorDataTypes::Int8 };
-const static SupportedTensorDataTypes supportedTypeListDequantizeLinear[2] = { SupportedTensorDataTypes::Float32, SupportedTensorDataTypes::UInt8 | SupportedTensorDataTypes::Int8 | SupportedTensorDataTypes::Int32 };
-const static SupportedTensorDataTypes supportedTypeListQuantize[2] = { SupportedTensorDataTypes::Float32, SupportedTensorDataTypes::UInt8 };
-const static SupportedTensorDataTypes supportedTypeListIsNan[2] = { SupportedTensorDataTypes::Float16to32, SupportedTensorDataTypes::Bool };
-const static SupportedTensorDataTypes supportedTypeListIsInf[2] = { SupportedTensorDataTypes::Float32, SupportedTensorDataTypes::Bool };
-const static SupportedTensorDataTypes supportedTypeListConstantOfShape[2] = { SupportedTensorDataTypes::Int32|SupportedTensorDataTypes::Int64, SupportedTensorDataTypes::Float16to32 };
-const static SupportedTensorDataTypes supportedTypeListWhere[2] = { SupportedTensorDataTypes::Bool, SupportedTensorDataTypes::AllScalars };
-const static SupportedTensorDataTypes supportedTypeListOneHot[3] = /* indices, depth, values */ { SupportedTensorDataTypes::Int32to64, SupportedTensorDataTypes::AllScalars, SupportedTensorDataTypes::Scalars8to32 };
-const static SupportedTensorDataTypes supportedTypeListLogicalComparison7[2] = /* A&B,C */ { SupportedTensorDataTypes::Float16to32, SupportedTensorDataTypes::Bool };
-const static SupportedTensorDataTypes supportedTypeListLogicalComparison9[2] = /* A&B,C */ { SupportedTensorDataTypes::NumericDefault, SupportedTensorDataTypes::Bool };
-const static SupportedTensorDataTypes supportedTypeListSigned[1] = { SupportedTensorDataTypes::Float16to32 | SupportedTensorDataTypes::Int32 | SupportedTensorDataTypes::Int16 | SupportedTensorDataTypes::Int8 };
-const static SupportedTensorDataTypes supportedTypeListRange[1] = {SupportedTensorDataTypes::Int16|SupportedTensorDataTypes::Int32|SupportedTensorDataTypes::Float32};
-const static SupportedTensorDataTypes supportedTypeListInteger[3] = {SupportedTensorDataTypes::Int8|SupportedTensorDataTypes::UInt8, SupportedTensorDataTypes::Int8|SupportedTensorDataTypes::UInt8, SupportedTensorDataTypes::Int32 };
-const static SupportedTensorDataTypes supportedTypeListPadWithoutFloat16[1] = { SupportedTensorDataTypes::Int8to32 | SupportedTensorDataTypes::Float32 };
-const static SupportedTensorDataTypes supportedTypeListQLinearMatMul[3] = {
+constexpr static std::array<const char*, 1> typeNameListDefault = {"T"};
```

Comment:
MSVC is giving me a compile error constructing the constexpr gsl::span from a constexpr C array. From span's source code it looks like it should work, but for some reason the compiler is inferring the array size incorrectly.

To repro:

* Use plain C array
* Register any operator w/ a plain array of size > 1, for example typeNameListThree: `constexpr OperatorRegistrationInformation this_breaks{REG_INFO(      9,  OneHot,                             typeNameListThree,              supportedTypeListOneHot,            DmGraphSupport::Supported,      requiredConstantCpuInputs(1))};`

Error message:

```
C:\Users\tiago\src\onnxruntime\onnxruntime\core\providers\dml\DmlExecutionProvider\src\Operators\OperatorRegistration.cpp(321): error C2131: expression did not evaluate to a constant
C:\Users\tiago\src\onnxruntime\onnxruntime\gsl\gsl-lite.hpp(1628): note: failure was caused by out of range index 3; allowed range is 0 <= index < 1
```

gsl-lite.hpp:1628 is this constructor:

```
template <class U, size_t N>
    gsl_api gsl_constexpr span(U (&arr)[N]) gsl_noexcept
        : first_(gsl_ADDRESSOF(arr[0])),
          last_(gsl_ADDRESSOF(arr[0]) + N) {}
```

Reply:
Sigh, oh well. Maybe ORT uses an older version of GSL, because it worked for me locally in WindowsAI.

Reply:
ORT is using GSL-lite to be compatible with pre-C++-14 compilers.

Reply:
By the way, in case you're curious about this, it is indeed a MSVC bug that makes it reject or crash on some valid pointer arithmetic in constexpr. I've hit this in other constexpr expressions and opened a bug here: https://developercommunity.visualstudio.com/content/problem/1110064/valid-pointer-arithmetic-rejected-in-constexpr-con.html